src/net/proxy.rs

```diff
@@ -0,0 +1,376 @@
+//! # Proxy support.
+//!
+//! Delta Chat supports SOCKS5 and shadowsocks protocols.
+
+use std::fmt;
+use std::pin::Pin;
+
+use anyhow::{format_err, Context as _, Result};
+use fast_socks5::client::Socks5Stream;
+use fast_socks5::util::target_addr::ToTargetAddr;
+use fast_socks5::AuthenticationMethod;
+use fast_socks5::Socks5Command;
+use percent_encoding::{percent_encode, NON_ALPHANUMERIC};
+use pin_project::pin_project;
+use tokio::io::{AsyncRead, AsyncWrite};
+use tokio::net::TcpStream;
+use tokio_io_timeout::TimeoutStream;
+
+use crate::context::Context;
+use crate::net::connect_tcp;
+use crate::net::session::SessionStream;
+use crate::sql::Sql;
+
+/// Default SOCKS5 port according to [RFC 1928](https://tools.ietf.org/html/rfc1928).
+pub const DEFAULT_SOCKS_PORT: u16 = 1080;
+
+#[derive(Debug, Clone)]
+pub struct ShadowsocksConfig {
+    pub server_config: shadowsocks::config::ServerConfig,
+}
+
+impl PartialEq for ShadowsocksConfig {
+    fn eq(&self, other: &Self) -> bool {
+        self.server_config.to_url() == other.server_config.to_url()
+    }
+}
+
+impl Eq for ShadowsocksConfig {}
+
+/// Wrapper for shadowsocks stream implementing `Debug`.
```

Comment:
Would the approach described in https://stackoverflow.com/a/65366047 (see `DimsWrapper`) fit here? Or AsyncRead/AsyncWrite still would need to be implemented?